src/openzeppelin/upgrades/upgradeable.cairo

```diff
@@ -0,0 +1,54 @@
+#[contract]
+mod Upgradeable {
+    use starknet::class_hash::ClassHash;
+    use starknet::class_hash::ClassHashZeroable;
+    use starknet::ContractAddress;
+    use starknet::ContractAddressZeroable;
+    use starknet::get_caller_address;
+    use starknet::syscalls::replace_class_syscall;
+    use zeroable::Zeroable;
+
+    struct Storage {
+        admin: ContractAddress,
+        initialized: bool,
+    }
+
+    #[event]
+    fn Upgraded(implementation: ClassHash) {}
+
+    #[event]
+    fn AdminChanged(previous_admin: ContractAddress, new_admin: ContractAddress) {}
+
+    fn initializer(contract_admin: ContractAddress) {
+        assert(!initialized::read(), 'Contract already initialized');
+        initialized::write(true);
+        _set_admin(contract_admin);
```

Comment:
i'm seriously questioning the need for an enshrined proxy admin now that there's no proxy. maybe the Upgrades lib should just offer the `_upgrade` method with the proper checks (e.g. disallow `class_hash == 0`, and let the users provide their own access control through the homonymous module or ownable